src/System.Private.CoreLib/src/System/MulticastDelegate.cs

```diff
@@ -572,7 +573,7 @@ private void CtorClosed(object target, IntPtr methodPtr)
         {
             if (target == null)
                 ThrowNullThisInDelegateToInstance();
-            this._target = target;
+            this._target = target!;
```

Comment:
Should we add TODO tag to all these places where `!` is used just because of the throw helpers - so to clean them up later?

Reply:
Yes.  I've been adding "// TODO-NULLABILITY: explanation" in cases where we expect to be able to remove the `!` in the future.  I agree we should do it in these cases as well.

That said, I've also started using Debug.Asserts instead of `!` in cases where I don't think the compiler will ever be able to tell and we may as well get the extra checked validation.

Reply:
Is that level of inference contemplated?

As for `Debug.Assert`, do we really want to write code like this, nullability feature aside? It's clumsy to read - reminds me of native code sprinkled with null pointer checks.

```c#
            if (foo == null)
                ThrowSomething();

           Debug.Assert(foo != null);

            // .. continue as normal
```

Reply:
Ah, I see.